cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/ProgramLifecycleHttpHandler.java

```diff
@@ -305,12 +310,24 @@ public void performAction(HttpRequest request, HttpResponder responder,
       throw new BadRequestException(String.format("Unknown program type '%s'", type), e);
     }
 
-    if ("debug".equals(action) && !isDebugAllowed(programType)) {
-      throw new NotImplementedException(String.format("debug action is not implemented for program type %s",
-                                                      programType));
-    }
     ProgramId programId = Ids.namespace(namespaceId).app(appId).program(programType, id);
-    lifecycleService.startStopProgram(programId, decodeArguments(request), action);
+    Map<String, String> args = decodeArguments(request);
+    // we have already validated that the action is valid
+    switch (action.toLowerCase()) {
+      case "start":
+        lifecycleService.start(programId, args, false);
+        break;
+      case "debug":
+        if (!isDebugAllowed(programType)) {
+          throw new NotImplementedException(String.format("debug action is not implemented for program type %s",
+                                                          programType));
+        }
+        lifecycleService.start(programId, args, true);
+        break;
+      case "stop":
+        lifecycleService.stop(programId);
+        break;
+    }
```

Comment:
We can have `default` case here and throw an Exception and remove `isValidAction` method